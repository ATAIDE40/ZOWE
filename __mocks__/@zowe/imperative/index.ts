/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Mocks all imperative classes specified under __mocks__/__imperative__
 * @returns mocked version of the @zowe/imperative module
 */
const __imperative__get = function () {
  const fs = require.requireActual("fs");
  const imperativePath = require("path").resolve(__dirname, "..", "..", "__imperative__");
  const extraClasses: any = {};
  fs.readdirSync(imperativePath).forEach((file: string) => {
    extraClasses[file.split(".")[0]] = require(`${imperativePath}/${file}`);
  });
  return {
    ...require.requireActual("@zowe/imperative"),
    // Note: Comment out the line below if you do not want to mock
    //       the classes in __mocks__/__imperative__/ for local testing purposes
    ...extraClasses
    // Note: You can still use Imperative.console.info("debug") in your test
  };
}

module.exports = __imperative__get();
