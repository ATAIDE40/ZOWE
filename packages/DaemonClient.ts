import { Imperative } from "@zowe/imperative";
import * as net from "net";


/**
 * Class for handling client connections to our persistent service (e.g. daemon mode)
 * @export
 * @class DaemonClient
 */
export class DaemonClient {

    /**
     * Undocumented paramter to cause daemon to shutdown
     * @private
     * @static
     * @memberof DaemonClient
     */
    private static readonly STOP_KEY = "--shutdown";

    /**
     * Creates an instance of DaemonClient.
     * @param {net.Socket} mClient
     * @param {net.Server} mServer
     * @memberof DaemonClient
     */
    constructor(private mClient: net.Socket, private mServer: net.Server) {
    }

    /**
     * Run an instance of this client and wait for proper events
     * @memberof DaemonClient
     */
    public run() {
        Imperative.api.appLogger.trace('daemon client connected');
        this.mClient.on('end', this.end.bind(this));
        this.mClient.on('close', this.close.bind(this));
        this.mClient.on('data', this.data.bind(this));
    }

    /**
     * End event handler triggered when client disconnects
     * @private
     * @memberof DaemonClient
     */
    private end() {
        Imperative.api.appLogger.trace('daemon client disconnected');
    }

    /**
     * Close event handler triggered when client closes connection
     * @private
     * @memberof DaemonClient
     */
    private close() {
        Imperative.api.appLogger.trace('client closed');
    }

    /**
     * Data event handler triggered for whenever data comes in on a connection
     * @private
     * @param {Buffer} data
     * @memberof DaemonClient
     */
    private data(data: Buffer) {
        // NOTE(Kelosky): this is not exposed yet, but will allow for a clean shut down if undocumented `--shutdown`
        // is written to the persistent Processor.
        const stopOffset = data.toString().indexOf(DaemonClient.STOP_KEY);
        if (stopOffset > -1) {
            if (this.mServer) {
            Imperative.api.appLogger.debug("shutting down")
                this.mClient.write(`Terminating server`);
                this.mClient.end();
                this.mServer.close()
            }
            // accept input parameters as we do without running in a server mode and pass our clients stream
            // handle as context
        } else {
            Imperative.api.appLogger.trace(`daemon input command: ${data.toString()}`)
            Imperative.commandLine = data.toString();
            Imperative.parse(data.toString(), { stream: this.mClient });
        }
    }
}