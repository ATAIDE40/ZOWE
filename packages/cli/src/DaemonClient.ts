/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Imperative } from "@zowe/imperative";
import * as net from "net";


/**
 * Class for handling client connections to our persistent service (e.g. daemon mode)
 * @export
 * @class DaemonClient
 */
export class DaemonClient {

    /**
     * Undocumented paramter to cause daemon to shutdown
     * @private
     * @static
     * @memberof DaemonClient
     */
    private static readonly STOP_KEY = "--shutdown";
    private static readonly ZOWE_DEAMON_HEADER_KEY = "x-zowe-daemon";

    /**
     * Creates an instance of DaemonClient.
     * @param {net.Socket} mClient
     * @param {net.Server} mServer
     * @memberof DaemonClient
     */
    constructor(private mClient: net.Socket, private mServer: net.Server) {
    }

    /**
     * Run an instance of this client and wait for proper events
     * @memberof DaemonClient
     */
    public run() {
        Imperative.api.appLogger.trace('daemon client connected');
        this.mClient.on('end', this.end.bind(this));
        this.mClient.on('close', this.close.bind(this));
        this.mClient.on('data', this.data.bind(this));
    }

    /**
     * End event handler triggered when client disconnects
     * @private
     * @memberof DaemonClient
     */
    private end() {
        Imperative.api.appLogger.trace('daemon client disconnected');
    }

    /**
     * Close event handler triggered when client closes connection
     * @private
     * @memberof DaemonClient
     */
    private close() {
        Imperative.api.appLogger.trace('client closed');
    }

    /**
     * Data event handler triggered for whenever data comes in on a connection
     * @private
     * @param {Buffer} data
     * @memberof DaemonClient
     */
    private data(data: Buffer) {
        if (data.toString().indexOf(DaemonClient.ZOWE_DEAMON_HEADER_KEY) > -1) {
            return;
        }
        // NOTE(Kelosky): this is not exposed yet, but will allow for a clean shut down if undocumented `--shutdown`
        // is written to the persistent Processor.  It should be wrapped in a new header and handled in DaemonClient.ts, e.g. x-zowe-daemon-shutdown
        const stopOffset = data.toString().indexOf(DaemonClient.STOP_KEY);
        if (stopOffset > -1) {
            if (this.mServer) {
                Imperative.api.appLogger.debug("shutting down")
                this.mClient.write(`Terminating server`);
                this.mClient.end();
                this.mServer.close()
            }
            // accept input parameters as we do without running in a server mode and pass our clients stream
            // handle as context
        } else {
            Imperative.api.appLogger.trace(`daemon input command: ${data.toString()}`)
            Imperative.commandLine = data.toString();
            Imperative.parse(data.toString(), { stream: this.mClient });
        }
    }
}