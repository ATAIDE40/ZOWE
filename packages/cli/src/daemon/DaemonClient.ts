/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as net from "net";
import { Duplex, Readable } from "stream";
import { DaemonRequest, IDaemonResponse, Imperative, ImperativeError } from "@zowe/imperative";

/**
 * Class for handling client connections to our persistent service (e.g. daemon mode)
 * @export
 * @class DaemonClient
 */
export class DaemonClient {
    /**
     * The character sent when Ctrl+C is pressed to terminate a process.
     * @internal
     */
    public static readonly CTRL_C_CHAR = "\x03";

    /**
     * The number of stdin bytes remaining to read from the daemon client.
     */
    private stdinBytesRemaining = 0;

    /**
     * Creates an instance of DaemonClient.
     * @param {net.Socket} mClient
     * @param {net.Server} mServer
     * @param {string} mOwner
     * @memberof DaemonClient
     */
    constructor(private mClient: net.Socket, private mServer: net.Server, private mOwner: string) {
    }

    /**
     * Run an instance of this client and wait for proper events
     * @memberof DaemonClient
     */
    public run() {
        Imperative.api.appLogger.trace('daemon client connected');
        this.mClient.on('end', this.end.bind(this));
        this.mClient.on('close', this.close.bind(this));
        this.mClient.on('data', this.data.bind(this));
    }

    /**
     * End event handler triggered when client disconnects
     * @private
     * @memberof DaemonClient
     */
    private end() {
        Imperative.api.appLogger.trace('daemon client disconnected');
    }

    /**
     * Close event handler triggered when client closes connection
     * @private
     * @memberof DaemonClient
     */
    private close() {
        Imperative.api.appLogger.trace('client closed');
    }

    /**
     * Shutdown the daemon server cleanly
     * @private
     * @memberof DaemonClient
     */
    private shutdown() {
        // NOTE(Kelosky): this is not exposed yet, but will allow for a clean shut down if
        // undocumented `--shutdown` is written to the persistent Processor.  It should be wrapped
        // in a new header and handled in DaemonClient.ts, e.g. x-zowe-daemon-shutdown
        Imperative.api.appLogger.debug("shutting down");
        this.mClient.write(`Terminating server`);
        this.mClient.end();
        this.mServer.close();
    }

    /**
     * Write data received from the daemon client to stdin.
     * @param data Data to write to stdin
     * @param expectedLength Expected length of the data to validate
     * @private
     * @memberof DaemonClient
     */
    private createStdinStream(data: Buffer, expectedLength: number): Readable | undefined {
        if (data == null) return;

        const stream = new Duplex();
        stream.push(data);
        this.stdinBytesRemaining = expectedLength - data.byteLength;

        if (this.stdinBytesRemaining > 0) {
            const outer: DaemonClient = this;  // eslint-disable-line @typescript-eslint/no-this-alias
            this.mClient.on("data", function listener(data) {
                stream.push(data);
                this.stdinBytesRemaining -= data.byteLength;

                if (this.stdinBytesRemaining <= 0) {
                    stream.push(null);
                    outer.mClient.removeListener("data", listener);
                }
            });
        } else {
            stream.push(null);
        }

        return stream;
    }

    /**
     * Data event handler triggered for whenever data comes in on a connection
     * @private
     * @param {Buffer} data
     * @memberof DaemonClient
     */
    private async data(data: Buffer) {
        if (this.stdinBytesRemaining > 0) return;

        // Split JSON body and binary data from multipart response
        const jsonEndIdx = data.indexOf("}" + DaemonRequest.EOW_DELIMITER);
        let jsonData: IDaemonResponse;
        let stdinData: Buffer;

        try {
            jsonData = JSON.parse((jsonEndIdx !== -1 ? data.slice(0, jsonEndIdx + 1) : data).toString());
            stdinData = jsonEndIdx !== -1 ? data.slice(jsonEndIdx + 2) : undefined;
        } catch (error) {
            Imperative.api.appLogger.logError(new ImperativeError({
                msg: "Failed to parse data received from daemon client",
                causeErrors: error
            }));
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            Imperative.api.appLogger.trace("First 1024 bytes of daemon request:\n", data.slice(0, 1024).toString());
            const responsePayload: string = DaemonRequest.create({
                stderr: "Failed to parse data received from daemon client:\n" + error.stack,
                exitCode: 1
            });
            this.mClient.write(responsePayload);
            this.mClient.end();
            return;
        }

        let requestUser: string = undefined;
        if (jsonData.user != null) {
            try {
                requestUser = Buffer.from(jsonData.user, 'base64').toString();
            } catch (err) {
                Imperative.api.appLogger.error("The user field on a daemon request was malformed.");
            }
        }

        if (requestUser == null || requestUser === '') {
            // Someone tried connecting but is missing something important.
            Imperative.api.appLogger.warn("A connection was attempted without a valid user.");
            const responsePayload: string = DaemonRequest.create({
                stderr: "The daemon client did not supply user information or supplied bad information.\n",
                exitCode: 1
            });
            this.mClient.write(responsePayload);
            this.mClient.end();
            return;
        } else if (requestUser != this.mOwner) {
            // Someone else is trying to use the daemon, and should be stopped.
            Imperative.api.appLogger.warn("The user '" + requestUser + "' attempted to connect.");
            const responsePayload: string = DaemonRequest.create({
                stderr: "The user '" + requestUser + "' cannot use this daemon.\n",
                exitCode: 1
            });
            this.mClient.write(responsePayload);
            this.mClient.end();
            return;
        }

        if (jsonData.stdin != null) {
            if (jsonData.stdin !== DaemonClient.CTRL_C_CHAR) {
                // This data is related to a prompt reply so we ignore it
                return;
            } else if (this.mServer) {
                // Ctrl+C signal was sent so we shutdown the server
                this.shutdown();
            }
        } else {
            Imperative.commandLine = jsonData.argv.join(" ");
            Imperative.api.appLogger.trace(`daemon input command: ${Imperative.commandLine}`);
            Imperative.parse(jsonData.argv, {
                stream: this.mClient,
                stdinStream: this.createStdinStream(stdinData, jsonData.stdinLength),
                request: jsonData
            });
        }
    }
}
