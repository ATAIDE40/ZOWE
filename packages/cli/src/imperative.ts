/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";
import { ZosFilesOptions } from "./zosfiles/ZosFiles.options";

import { ZosmfSession } from "@zowe/zosmf-for-zowe-sdk";
import { ZosTsoProfile } from "@zowe/zos-tso-for-zowe-sdk";
import { ZosUssProfile } from "@zowe/zos-uss-for-zowe-sdk";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: [
        "daemon/*.definition!(.d).*s",
        "provisioning/*.definition!(.d).*s",
        "workflows/*.definition!(.d).*s",
        "zosconsole/*.definition!(.d).*s",
        "zosfiles/*.definition!(.d).*s",
        "zosjobs/*.definition!(.d).*s",
        "zosmf/*.definition!(.d).*s",
        "zostso/*.definition!(.d).*s",
        "zosuss/*.definition!(.d).*s",
        "zoslogs/*.definition!(.d).*s"
    ],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help-logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    apimlConnLookup: [
        {
            apiId: "ibm.zosmf",
            gatewayUrl: "api/v1",
            connProfType: "zosmf"
        }
    ],
    baseProfile: {
        type: "base",
        schema: {
            type: "object",
            title: "Base Profile",
            description: "Base profile that stores values shared by multiple service profiles",
            properties: {
                host: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_HOST,
                    includeInTemplate: true
                },
                port: {
                    type: "number",
                    optionDefinition: Constants.BASE_OPTION_PORT
                },
                user: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_USER,
                    includeInTemplate: true
                },
                password: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_PASSWORD,
                    includeInTemplate: true
                },
                rejectUnauthorized: {
                    type: "boolean",
                    optionDefinition: Constants.BASE_OPTION_REJECT_UNAUTHORIZED,
                    includeInTemplate: true
                },
                tokenType: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_TOKEN_TYPE
                },
                tokenValue: {
                    type: "string",
                    secure: true,
                    optionDefinition: Constants.BASE_OPTION_TOKEN_VALUE
                },
                certFile: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_CERT_FILE
                },
                certKeyFile: {
                    type: "string",
                    optionDefinition: Constants.BASE_OPTION_CERT_KEY_FILE
                // },
                // certFilePassphrase: {
                //     type: "string",
                //     secure: true,
                //     optionDefinition: Constants.BASE_OPTION_CERT_FILE_PASSPHRASE
                }
            },
            required: []
        },
        createProfileExamples: [
            {
                options: "base1 --host example.com --port 443 --user admin --password 123456",
                description: "Create a profile called 'base1' to connect to host example.com and port 443"
            },
            {
                options: "base2 --host example.com --user admin --password 123456 --reject-unauthorized false",
                description: "Create a profile called 'base2' to connect to host example.com (default port - 443) " +
                    "and allow self-signed certificates"
            },
            {
                options: "base3 --host example.com --port 1443",
                description: "Create a profile called 'base3' to connect to host example.com and port 1443, " +
                    " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
            },
            {
                options: "base4 --reject-unauthorized false",
                description: "Create a zosmf profile called 'base4' to connect to default port 443 and allow self-signed certificates, " +
                    "not specifying a username, password, or host so they are not stored on disk; these will need to be specified on every command"
            }
        ],
        updateProfileExamples: [
            {
                options: "base1 --user newuser --password newp4ss",
                description: "Update a base profile named 'base1' with a new username and password"
            }
        ],
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/ApimlAuthHandler",
                login: {
                    summary: Constants.APIML_LOGIN_SUMMARY,
                    description: Constants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGIN_EXAMPLE1,
                        Constants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.BASE_OPTION_USER,
                        Constants.BASE_OPTION_PASSWORD,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED,
                        Constants.BASE_OPTION_CERT_FILE,
                        Constants.BASE_OPTION_CERT_KEY_FILE
                    ]
                },
                logout: {
                    summary: Constants.APIML_LOGOUT_SUMMARY,
                    description: Constants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        Constants.APIML_LOGOUT_EXAMPLE1,
                        Constants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        Constants.BASE_OPTION_HOST,
                        Constants.BASE_OPTION_PORT,
                        Constants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        Constants.BASE_OPTION_TOKEN_VALUE,
                        Constants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: Constants.AUTH_GROUP_SUMMARY,
            description: Constants.AUTH_GROUP_DESCRIPTION
        }
    },
    configAutoInitCommandConfig: {
        handler: __dirname + "/config/auto-init/ApimlAutoInitHandler",
        provider: "APIML",
        autoInit: {
            options: [
                Constants.AUTO_INIT_OPTION_HOST,
                Constants.AUTO_INIT_OPTION_PORT,
                Constants.AUTO_INIT_OPTION_USER,
                Constants.AUTO_INIT_OPTION_PASSWORD,
                Constants.AUTO_INIT_OPTION_REJECT_UNAUTHORIZED,
                Constants.AUTO_INIT_OPTION_TOKEN_TYPE,
                Constants.AUTO_INIT_OPTION_TOKEN_VALUE
            ]
        },
        profileType: "base"
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_HOST
                    },
                    port: {
                        type: "number",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PORT,
                        includeInTemplate: true
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_USER
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PASSWORD
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED
                    },
                    certFile: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_FILE
                    },
                    certKeyFile: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_KEY_FILE
                    },
                    // certFilePassphrase: {
                    //     type: "string",
                    //     secure: true,
                    //     optionDefinition: ZosmfSession.ZOSMF_OPTION_CERT_FILE_PASSPHRASE
                    // },
                    basePath: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_BASE_PATH
                    },
                    protocol: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PROTOCOL
                    },
                    encoding: {
                        type: "string",
                        optionDefinition: {
                            name: "encoding",
                            aliases: ["ec"],
                            description: "The encoding for download and upload of z/OS data set and USS files." +
                                " The default encoding if not specified is IBM-1047.",
                            type: "string"
                        }
                    },
                    responseTimeout: {
                        type: "number",
                        optionDefinition: ZosFilesOptions.responseTimeout
                    }
                },
                required: []
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                },
                {
                    options: "zos125 --host zos125 --port 1443",
                    description: "Create a zosmf profile called 'zos125' to connect to z/OSMF at the host zos125 and port 1443, " +
                        " not specifying a username or password so they are not stored on disk; these will need to be specified on every command"
                },
                {
                    options: "zos126 --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos126' to connect to z/OSMF on the default port 443 and allow self-signed " +
                        "certificates, not specifying a username, password, or host so they are not stored on disk; these will need to be " +
                        "specified on every command"
                },
                {
                    options: "zosAPIML --host zosAPIML --port 2020 --user ibmuser --password myp4ss --reject-unauthorized false --base-path basePath",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                }
            ],
            updateProfileExamples: [
                {
                    options: "zos123 --user newuser --password newp4ss",
                    description: "Update a zosmf profile named 'zos123' with a new username and password"
                }
            ]
        },
        ZosTsoProfile,
        ZosUssProfile
    ]
};
module.exports = config;
