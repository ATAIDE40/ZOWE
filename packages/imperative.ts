/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

// WIP Imperative version of Zowe CLI
import { IImperativeConfig } from "@brightside/imperative";
import { Constants } from "./Constants";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: ["**/cli/*.definition!(.d).*s"],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        },
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["H"],
                            description: "The z/OSMF server host name.",
                            type: "string",
                            required: true,
                        },
                    },
                    port: {
                        type: "number",
                        optionDefinition: {
                            name: "port",
                            aliases: ["P"],
                            description: "The z/OSMF server port.",
                            type: "number",
                            defaultValue: 443,
                        },
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "user",
                            aliases: ["u"],
                            description: "Mainframe (z/OSMF) user name, which can be the same as your TSO login.",
                            type: "string",
                            implies: ["password"],
                        },
                    },
                    pass: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "password",
                            aliases: ["p"],
                            description: "Mainframe (z/OSMF) password, which can be the same as your TSO password.",
                            type: "string",
                            implies: ["user"],
                        },
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: {
                            name: "reject-unauthorized",
                            aliases: ["ru"],
                            description: "Reject self-signed certificates.",
                            type: "boolean",
                            defaultValue: "true",
                        },
                    },
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                    "and allow self-signed certificates"
                }
            ]
        },
        {
            type: "tso",
            schema: {
                type: "object",
                title: "TSO Profile",
                description: "z/OS TSO/E User Profile",
                properties: {
                    account: {
                        type: "string",
                        optionDefinition: {
                            name: "account",
                            aliases: ["a"],
                            description: "Your z/OS TSO/E accounting information.",
                            type: "string",
                            required: true,
                        },
                    },
                    characterSet: {
                        type: "string",
                        optionDefinition: {
                            name: "character-set",
                            aliases: ["cs"],
                            description: "Character set for address space to convert messages and responses from UTF-8 to EBCDIC.",
                            type: "string",
                            defaultValue: "697",
                        },
                    },
                    codePage: {
                        type: "string",
                        optionDefinition: {
                            name: "code-page",
                            aliases: ["cp"],
                            description: "Codepage value for TSO/E address space to convert messages and responses from UTF-8 to EBCDIC.",
                            type: "string",
                            defaultValue: "1047",
                        },
                    },
                    columns: {
                        type: "number",
                        optionDefinition: {
                            name: "columns",
                            aliases: ["cols"],
                            description: "The number of columns on a screen.",
                            type: "number",
                            defaultValue: 80,
                        },
                    },
                    logonProcedure: {
                        type: "string",
                        optionDefinition: {
                            name: "logon-procedure",
                            aliases: ["l"],
                            description: "The logon procedure to use when creating TSO procedures on your behalf.",
                            type: "string",
                            defaultValue: "IZUFPROC",
                        },
                    },
                    regionSize: {
                        type: "number",
                        optionDefinition: {
                            name: "region-size",
                            aliases: ["rs"],
                            description: "Region size for the TSO/E address space.",
                            type: "number",
                            defaultValue: 4096,
                        },
                    },
                    rows: {
                        type: "number",
                        optionDefinition: {
                            name: "rows",
                            description: "The number of rows on a screen.",
                            type: "number",
                            defaultValue: 24,
                        },
                    },
                },
                required: ["account"],
            },
            createProfileExamples: [
                {
                    description: "Create a tso profile called 'myprof' with default settings and JES accounting information of 'IZUACCT'",
                    options: "myprof -a IZUACCT"
                },
                {
                    description: "Create a tso profile called 'largeregion' with a region size of 8192, a logon procedure of MYPROC, and " +
                    "JES accounting information of '1234'",
                    options: "largeregion -a 1234 --rs 8192"
                }
            ]
        },
    ]
};
module.exports = config;
