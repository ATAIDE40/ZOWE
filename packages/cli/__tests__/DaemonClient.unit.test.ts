/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

jest.mock("net");
jest.mock("@zowe/imperative");
import * as net from "net";
import { Imperative } from "@zowe/imperative";
import { DaemonClient } from "../src/DaemonClient";

describe("DaemonClient tests", () => {

    it("should process data when received", () => {

        const log = jest.fn(() => {
            // do nothing
        });

        const parse = jest.fn( (data, context) => {
            expect(data).toMatchSnapshot();
            expect(context).toMatchSnapshot();
        });

        (Imperative as any) = {
            api: {
                appLogger: {
                    trace: log,
                    debug: log
                }
            },
            commandLine: "n/a",
            parse
        };

        const server: net.Server = undefined;
        const client = {
            on: jest.fn()
        };

        const daemonClient = new DaemonClient(client as any, server);

        daemonClient.run();
        // force `data` call and verify input is from instantiation of DaemonClient
        // and is what is passed to mocked Imperative.parse via snapshot
        (daemonClient as any).data("PWD\fsome data", {whatever: "context I want"});
    });

    it("should ignore JSON response data used for prompting when received", () => {

        const log = jest.fn(() => {
            // do nothing
        });

        const parse = jest.fn( (data, context) => {
            // do nothing
        });

        (Imperative as any) = {
            api: {
                appLogger: {
                    trace: log,
                    debug: log
                }
            },
            commandLine: "n/a",
            // parse
        };

        const server: net.Server = undefined;
        const client = {
            on: jest.fn()
        };

        const daemonClient = new DaemonClient(client as any, server);

        daemonClient.run();
        // force `data` call and verify input is from instantiation of DaemonClient
        // and is what is passed to mocked Imperative.parse via snapshot
        const promptResponse = { id: "daemon-client", reply: "some answer" };
        (daemonClient as any).data(JSON.stringify(promptResponse), {whatever: "context I want"});

        expect(parse).not.toHaveBeenCalled();
    });

    it("should shutdown when keyword is specified", () => {

        const log = jest.fn(() => {
            // do nothing
        });

        const parse = jest.fn( (data, context) => {
            expect(data).toMatchSnapshot();
            expect(context).toMatchSnapshot();
        });

        (Imperative as any) = {
            api: {
                appLogger: {
                    trace: log,
                    debug: log
                }
            },
            commandLine: "n/a",
            parse
        };

        const server = {
            close: jest.fn()
        };
        const write = jest.fn((someWriteMessage) => {
            expect(someWriteMessage).toMatchSnapshot();
        });
        const client = {
            on: jest.fn(),
            write,
            end: jest.fn()
        };

        const daemonClient = new DaemonClient(client as any, server as any);
        daemonClient.run();
        // force `data` call and verify write method is called with termination message
        (daemonClient as any).data(Buffer.from("--shutdown"), {whatever: "context I want"});
    });

    it("should call the end method", () => {

        const log = jest.fn(() => {
            // do nothing
        });

        (Imperative as any).api = {
            appLogger: {
                trace: log
            }
        };

        const server: net.Server = undefined;
        const client = {
            on: jest.fn()
        };

        const daemonClient = new DaemonClient(client as any, server);
        daemonClient.run();
        (daemonClient as any).end();
        expect(log).toHaveBeenLastCalledWith('daemon client disconnected');
    });

    it("should call the close method", () => {

        const log = jest.fn(() => {
            // do nothing
        });

        (Imperative as any).api = {
            appLogger: {
                trace: log
            }
        };

        const server: net.Server = undefined;
        const client = {
            on: jest.fn()
        };

        const daemonClient = new DaemonClient(client as any, server);
        daemonClient.run();
        (daemonClient as any).close();
        expect(log).toHaveBeenLastCalledWith('client closed');
    });

});
