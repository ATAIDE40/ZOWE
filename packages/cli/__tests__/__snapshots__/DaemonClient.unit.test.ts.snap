// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DaemonClient tests should process data when received 1`] = `
Array [
  "feed",
  "dog",
]
`;

exports[`DaemonClient tests should process data when received 2`] = `
Object {
  "daemonResponse": Object {
    "argv": Array [
      "feed",
      "dog",
    ],
    "cwd": "fake",
    "env": Object {},
    "stdin": null,
    "stdinLength": 0,
  },
  "stream": Object {
    "on": [MockFunction] {
      "calls": Array [
        Array [
          "end",
          [Function],
        ],
        Array [
          "close",
          [Function],
        ],
        Array [
          "data",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
}
`;

exports[`DaemonClient tests should process response with binary stdin data when received 1`] = `
Array [
  "feed",
  "cat",
]
`;

exports[`DaemonClient tests should process response with binary stdin data when received 2`] = `
Object {
  "daemonResponse": Object {
    "argv": Array [
      "feed",
      "cat",
    ],
    "cwd": "fake",
    "env": Object {},
    "stdin": null,
    "stdinLength": 11,
  },
  "stream": Object {
    "on": [MockFunction] {
      "calls": Array [
        Array [
          "end",
          [Function],
        ],
        Array [
          "close",
          [Function],
        ],
        Array [
          "data",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
}
`;

exports[`DaemonClient tests should shutdown when keyword is specified 1`] = `"Terminating server"`;
